File: cleanup_episodes/lib/cleanup-episodes/handler.rb
require "aws-sdk-s3"
require "yaml"
require "json"
require "date"

module Radicaster
  module CleanupEpisodes
    # m4aファイルアップロード時に発動し、保持ポリシーに基づいて古いエピソードを削除する
    class Handler
      EPISODE_EXT = ".m4a"
      # ファイル名から日付を抽出する正規表現 (YYYYMMDD.m4a)
      DATE_PATTERN = /(\d{8})\.m4a\z/

      def initialize(logger, s3_client, bucket)
        @logger = logger
        @s3_client = s3_client
        @bucket = bucket
      end

      def handle(event:, context:)
        logger.info("Start cleanup episode handling.")
        id = extract_id(event)
        definition = load_definition(id)

        unless definition
          logger.info("Definition not found for id: #{id}. Skipping.")
          return
        end

        retention_type = definition["retention_type"]
        retention_value = definition["retention_value"]

        unless retention_type && retention_value
          logger.info("No retention policy set for id: #{id}. Skipping.")
          return
        end

        case retention_type
        when "count"
          cleanup_by_count(id, retention_value.to_i)
        when "days"
          cleanup_by_days(id, retention_value.to_i)
        else
          logger.info("Unknown retention_type: #{retention_type}. Skipping.")
        end

        logger.info("Cleanup episode handling finished.")
      end

      private

      attr_reader :logger, :s3_client, :bucket

      def extract_id(event)
        event = JSON.parse(event["Records"][0]["Sns"]["Message"]) if event["Records"][0]["Sns"]
        raise '"s3" is not contained in the event' unless event["Records"][0]["s3"]
        key = event["Records"][0]["s3"]["object"]["key"]
        logger.debug("S3 key: #{key}")
        # key format: {id}/data/{YYYYMMDD}.m4a
        key.split("/").first
      end

      def load_definition(id)
        key = "radicaster/#{id}.yaml"
        resp = s3_client.get_object(bucket: bucket, key: key)
        YAML.load(resp.body.read)
      rescue Aws::S3::Errors::NoSuchKey
        nil
      end

      def list_episodes(id)
        prefix = "#{id}/data/"
        resp = s3_client.list_objects_v2(bucket: bucket, prefix: prefix)
        return [] unless resp.contents

        resp.contents
          .select { |obj| obj.key.end_with?(EPISODE_EXT) }
          .sort_by { |obj| obj.key }
      end

      def extract_date_from_key(key)
        match = key.match(DATE_PATTERN)
        return nil unless match
        Date.strptime(match[1], "%Y%m%d")
      rescue Date::Error
        nil
      end

      def cleanup_by_count(id, keep_count)
        episodes = list_episodes(id)
        logger.info("Found #{episodes.size} episodes for id: #{id}, keeping last #{keep_count}")

        return if episodes.size <= keep_count

        # ファイル名(日付)でソートされているので、古い順に並んでいる
        # 末尾のkeep_count個を残して、先頭の古いものを削除
        to_delete = episodes[0..-(keep_count + 1)]

        to_delete.each do |obj|
          logger.info("Deleting old episode: #{obj.key}")
          s3_client.delete_object(bucket: bucket, key: obj.key)
        end

        logger.info("Deleted #{to_delete.size} episodes for id: #{id}")
      end

      def cleanup_by_days(id, keep_days)
        episodes = list_episodes(id)
        cutoff_date = Date.today - keep_days
        logger.info("Found #{episodes.size} episodes for id: #{id}, deleting before #{cutoff_date}")

        deleted_count = 0
        episodes.each do |obj|
          episode_date = extract_date_from_key(obj.key)
          next unless episode_date

          if episode_date < cutoff_date
            logger.info("Deleting old episode: #{obj.key} (date: #{episode_date})")
            s3_client.delete_object(bucket: bucket, key: obj.key)
            deleted_count += 1
          end
        end

        logger.info("Deleted #{deleted_count} episodes for id: #{id}")
      end
    end
  end
end

File: cleanup_episodes/src/handler.ts
import { S3Client } from '@aws-sdk/client-s3';
import { S3Storage } from './s3';

const s3Client = new S3Client({});
const bucket = process.env.RADICASTER_S3_BUCKET!;

const storage = new S3Storage(s3Client, bucket);

// ファイル名から日付を抽出する正規表現 (YYYYMMDD.m4a)
const DATE_PATTERN = /(\d{8})\.m4a$/i;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function extractId(event: any): string {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let records: any[] = event.Records;

  // SNSイベントにS3イベントがラップされている場合はアンラップする
  if (records[0].Sns) {
    const snsMessage = JSON.parse(records[0].Sns.Message as string);
    records = snsMessage.Records;
  }

  const s3Record = records[0];
  if (!s3Record.s3) {
    throw new Error('"s3" is not contained in the event');
  }

  // S3イベントのキーはURLエンコードされている
  const key = decodeURIComponent((s3Record.s3.object.key as string).replace(/\+/g, ' '));
  // key format: {id}/data/{YYYYMMDD}.m4a
  return key.split('/')[0];
}

function extractDate(key: string): Date | null {
  const match = key.match(DATE_PATTERN);
  if (!match) return null;
  const dateStr = match[1];
  const year = parseInt(dateStr.slice(0, 4), 10);
  const month = parseInt(dateStr.slice(4, 6), 10) - 1;
  const day = parseInt(dateStr.slice(6, 8), 10);
  const date = new Date(year, month, day);
  if (isNaN(date.getTime())) return null;
  return date;
}

async function cleanupByCount(id: string, keepCount: number): Promise<void> {
  const episodes = await storage.listEpisodes(id);
  console.log(`Found ${episodes.length} episodes for id: ${id}, keeping last ${keepCount}`);

  if (episodes.length <= keepCount) return;

  // ファイル名（日付）昇順にソートされているので先頭の古いものを削除する
  const toDelete = episodes.slice(0, episodes.length - keepCount);
  for (const ep of toDelete) {
    console.log(`Deleting old episode: ${ep.key}`);
    await storage.deleteEpisode(ep.key);
  }
  console.log(`Deleted ${toDelete.length} episodes for id: ${id}`);
}

async function cleanupByDays(id: string, keepDays: number): Promise<void> {
  const episodes = await storage.listEpisodes(id);
  const cutoff = new Date();
  cutoff.setDate(cutoff.getDate() - keepDays);
  cutoff.setHours(0, 0, 0, 0);
  console.log(
    `Found ${episodes.length} episodes for id: ${id}, deleting before ${cutoff.toISOString().slice(0, 10)}`,
  );

  let deletedCount = 0;
  for (const ep of episodes) {
    const epDate = extractDate(ep.key);
    if (!epDate) continue;

    if (epDate < cutoff) {
      console.log(`Deleting old episode: ${ep.key} (date: ${epDate.toISOString().slice(0, 10)})`);
      await storage.deleteEpisode(ep.key);
      deletedCount++;
    }
  }
  console.log(`Deleted ${deletedCount} episodes for id: ${id}`);
}

/** SNSイベント（S3アップロード通知）をハンドルして古いエピソードを削除する */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export async function handler(event: any): Promise<void> {
  console.log('Start cleanup episode handling.');

  const id = extractId(event);
  console.debug(`id: ${id}`);

  const definition = await storage.findDefinition(id);
  if (!definition) {
    console.log(`Definition not found for id: ${id}. Skipping.`);
    return;
  }

  const retentionType = definition.retention_type;
  const retentionValue = definition.retention_value;

  if (!retentionType || retentionValue === undefined) {
    console.log(`No retention policy set for id: ${id}. Skipping.`);
    return;
  }

  if (retentionType === 'count') {
    await cleanupByCount(id, retentionValue);
  } else if (retentionType === 'days') {
    await cleanupByDays(id, retentionValue);
  } else {
    console.log(`Unknown retention_type: ${retentionType}. Skipping.`);
  }

  console.log('Cleanup episode handling finished.');
}
